src/ripple/overlay/impl/MessageScheduler.cpp

```diff
@@ -0,0 +1,413 @@
+                try
+                {
+                    it->second->receiver->onFailure(
+                        it->second->id, FailureCode::DISCONNECT);
+                }
+                catch (std::exception const&)
+                {
+                }
+                it = requests_.erase(it);
+            }
+            else
+            {
+                ++it;
+            }
+        }
+    }
+    if (!senders.empty())
+    {
+        if (!channels_.empty())
+        {
+            _offer(channels_, senders);
+        }
+        std::move(senders.begin(), senders.end(), std::back_inserter(senders_));
+    }
+}
+
+bool
+MessageScheduler::schedule(Sender* sender)
+{
+    JLOG(journal_.trace()) << "schedule,during=" << caller;
+    if (tsenders)
+    {
+        // The scheduler is already locked in this thread.
+        // Save new senders to be served later.
+        // TODO: Insert in priority order.
+        tsenders->emplace_back(sender);
+        return true;
+    }
+    std::lock_guard<std::mutex> lock(offersMutex_);
+    if (stopped_)
+    {
+        return false;
+    }
+    std::vector<Sender*> senders = {sender};
+    if (!channels_.empty())
+    {
+        push_value top(caller, "schedule");
+        _offer(channels_, senders);
+    }
+    std::move(senders.begin(), senders.end(), std::back_inserter(senders_));
+    return true;
+}
+
+RequestId
+MessageScheduler::send(
+    std::shared_ptr<Peer> peer,
+    protocol::TMGetLedger& message,
+    Receiver* receiver,
+    NetClock::duration timeout)
+{
+    auto requestId = ++nextId_;
+    message.set_requestcookie(requestId);
+    _send(peer, requestId, message, receiver, timeout);
+    return requestId;
+}
+
+RequestId
+MessageScheduler::send(
+    std::shared_ptr<Peer> peer,
+    protocol::TMGetObjectByHash& message,
+    Receiver* receiver,
+    NetClock::duration timeout)
+{
+    auto requestId = ++nextId_;
+    message.set_seq(requestId);
+    JLOG(journal_.trace()) << "send,type=get_objects,count="
+                           << message.objects_size();
+    _send(peer, requestId, message, receiver, timeout);
+    return requestId;
+}
+
+void
+MessageScheduler::_offer(
+    std::vector<Channel>& channels,
+    std::vector<Sender*>& senders)
+{
+    JLOG(journal_.trace()) << "offer,during=" << caller
+                           << ",channels=" << channels.size()
+                           << ",senders=" << senders.size();
+    assert(!channels.empty());
+    assert(!senders.empty());
+    push_value top1(tsenders, &senders);
+    // We must iterate `senders` using an index
+    // because the offer may grow the vector,
+    // invalidating any iterators or references.
+    for (auto i = 0; i < senders.size(); ++i)
+    {
+        // If this is the last sender, offer it the full set of channels.
+        // If there are more senders waiting, offer one at a time, in turn.
+        auto size = (i + 1 == senders.size()) ? channels.size() : 1;
```

Comment:
Since senders are in priority order, does this mean that a lower priority sender can get more channels than a higher priority sender? Could we offer the "top" sender more channels? Maybe reserve one channel for every sender, and offer the "top" sender all the remaining channels, instead of offering the last sender all the remaining channels?

Reply:
Right, this is an open question. The prioritization strategy in there now is basically a placeholder for something more sophisticated. How should we think about it? Here is the design I was thinking of:

- We want to allocate limited resource-time fairly among different callers. The resources here are channels, i.e. threads on peers. 
- Some callers may schedule multiple `Sender` objects, while others schedule only one. This should not let callers skirt their fair allocation. Multiple `Sender` objects from the same underlying caller should share that caller's quota.
- Each `Sender` and `Receiver` object gets a `Quota` object that identifies the caller. If `Sender`s were jobs, then `Quota`s would be job types. Each `Quota` has an integer weight, `w_q`, that represents its priority. This weight determines what portion of a given window of T resource-seconds should be allocated to that `Quota`'s requests. Call the sum of the weights of all outstanding quotas `W`.
- Whenever a `Receiver` is called (i.e. when a response is received or a timeout reached), insert an interval (send time, receive time) for the receiver's `Quota` into a queue. This interval is the **consumption** of resource-time.
- Whenever a channel opens, we need to decide the order in which to offer it to waiting senders. Iterate the consumption queue and calculate the sum total of resource-seconds consumed by each `Quota` within the most recent T seconds (i.e. ignore the parts of the intervals older than T seconds). For a `Quota` `q`, call this `c_q`. Call the sum of all `c_q`, `C`. Remove intervals from the beginning of the queue whose receive time is older than T seconds. These calculations can be done incrementally, updating totals based on new consumptions added and old ones removed, instead of iterating the full consumption queue each time.
- Then calculate the ratio `(c_q / C) / (w_q / W)` for each quota in the sender queue. Waiting senders are offered the available channels in the order of this ratio, lowest to highest.

Reply:
After refreshing myself more deeply, I remember now why this iteration was implemented. The first sender is offered 1 channel. If it doesn't close that channel, then it keeps its spot in the queue. If it closes that channel, then it is removed from the senders queue. If it needs more channels, then it can schedule another sender, which appears at the end of the queue. Then the next sender is offered 1 channel, and so on. Once we get to the end of the queue, that is the last sender we need to serve. Every prior sender has either closed all the channels it needed (and is thus removed from the queue) or passed. The final sender gets all the remaining channels because there are no more senders to divvy them among.